### ISSUE TITLE
Feature request: Enable Spark Support

### ISSUE THREAD

**Issue by ProbStub** [robertmartin8/PyPortfolioOpt#381]
**Is your feature request related to a problem?**
I like to perform large portfolio/a large number of asset optimizations, e.g. more than 1m portfolios and use Apache Spark rather than Pandas/Numpy.

**Describe the feature you'd like**
Large asset covariance computations should leverage parallel structures and portfolio optimization should be possible to scale to a larger number of portfolios. Similar to the [mocked up implementation on this fork](https://github.com/ProbStub/PyPortfolioOpt/tree/spark_mvo_ps_20211011).

**Additional context**
Hi,

I am currently thinking about using pypfopt in a spark environment. I have mocked up my idea on a fork [here](https://github.com/ProbStub/PyPortfolioOpt/tree/spark_mvo_ps_20211011) - it's work in progress. API and other changes are purely for demonstration purpose and clearly would need much more discussion if a PR were to be considered - also [spark 3.2’s pyspark.pandas](https://spark.apache.org/docs/3.2.0/api/python/getting_started/quickstart_df.html) will simplify the effort.

Before going ahead further I would have two questions:

1. Is spark support generally desired for pypfopt?
2. If yes, should that be an extension of the API as in the mock, or should it be a more stand-alone kind of implementation?

Cheers and happily looking forward to your thoughts.

Best regards,
Prob

**Comment by phschiele** [comment]
@ProbStub Very cool feature request! 

You mention scaling across many portfolios and assets, which to me are two quite different problems. The former can anyways be easily parallelized in most cases. For the latter, I'm really curious to hear what use cases you have in mind that require e.g. spark to compute.

**Comment by ProbStub** [comment]
@phschiele Yes two very different use cases. A large number of portfolios may result from the number of users and would, on its own not be a challange. However, the system would work with look-through ETF decomposition and re-compose the portfolios given a selection of metrics. An ETF portfolio may contain several positions of different share classes and even after aggregating up to the same securities the number of positions may be quite large. Hope that adds a little color.

**Comment by phschiele** [comment]
In my experience, the optimization usually would become a bottleneck first. 

With that said, it's cool that the koalas project will directly be included into PySpark, thanks for sharing that!

**Comment by blair-anson** [comment]
@ProbStub did you go any further with this?

**Comment by ProbStub** [comment]
@robertmartin8 I have had to stop working on this due to other commitments and the limited interest. Happy to pick it up again. There have been a few Spark releases since, so chances are this might be easier now. Back then then pandas cov function was still missing on the Pandas API for Spark.
